### PR TITLE
Tommasini Code Review by Tonin

### PR DESCRIPTION
**Pull request by MarcoTomma**
Hi! There you are the code for the last review! These are 5 files, actually they are 7 but you can only review those 5 where I'm the author, the file `SMSBackgroudHandler` is just to let you understand better the structure of the code ad how classes interact each other.

Let me briefly explain what's appening: `MainActivity` manages the front-end of this app, allowing the user to send a test message and also to receive an sms which will trigger an alarm and a notification.
The notification allows to stop the alarm or first opens the app and then stops it through a dialog box.

The tricky part is when we need to receive messages and launch the notification even on app closed: we must keep the broadcast receiver (which sadly is not there, it simply parse the sms and send it to `MessageReceivedService`), this was done creating a service `SMSKeepAliveService` that registers and keeps alive the broadcast receiver.

`MainActivityHelper` was written just for future uses and keeps the state of `MainActivity`.

For any question let me a comment, I hope you'll enjoy it.

### CODE REVIEW

**Review by tonin-alessandra — APPROVED** [review]
Don't be scared by my comments, they're almost all about possible hardwiring or suggestions about specs.
First of all, if I can afford to make a small criticism, this code is quite old, it dates back to one of the first sprints at the beginning of the course (I think about at the end of November), you only fixed some specifications now.
In any case, my task is to review it, so let's start with the last code review!
In general, the code is not so complicated, except for the SMSKeepAliveService, which could be a bit tricky, as you said.
The overall structure is good, the code is well-divided into different classes and methods, which handles different functions. Methods are not too long and they have the right number of parameters.
The code is well-formatted, names' conventions are respected (for example camelCase), variables/methods/classes' names are explicative and clear, and all this makes the code easy to read.
Maybe there's a bit of hardwiring sometimes, I pointed it out on my comments, but nothing particular.
So, good code overall, quite complete documentation and readable structure.
Nice job!

File: app/src/main/java/com/gruppo4/SMSApp/MainActivity.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            requestPermissions(new String[]{Manifest.permission.RECEIVE_SMS, Manifest.permission.SEND_SMS}, 1);
+    }
+
+    /**
+     * Checks SEND_SMS and RECEIVE_SMS permissions
+     * @return  true if the app has permissions, false otherwise
+     */
+    public boolean checkPermission() {
+        if (!(getApplicationContext().checkSelfPermission(Manifest.permission.SEND_SMS) == PackageManager.PERMISSION_GRANTED) ||
+                !(getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED))
```
[the remark this answers is not in the document]
Comment (MarcoTomma) [reply app/src/main/java/com/gruppo4/SMSApp/MainActivity.java]: You're right, I'll fix it instantly!

File: app/src/main/java/com/gruppo4/SMSApp/MainActivity.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    /**
+     * Creates and shows AlertDialog with two options:
+     * [stop] --> stops the ringtone and cancels the notification
+     * touch outside the dialog --> cancels the dialog and lets the ringtone ends by itself after a timeout
+     */
+    private void createStopRingDialog() {
+        AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
+        builder.setMessage("Your phone is ringing, stop it from here if you want");
```
[the remark this answers is not in the document]
Comment (MarcoTomma) [reply app/src/main/java/com/gruppo4/SMSApp/MainActivity.java]: I don't know how I didn't see that!

File: app/src/main/java/com/gruppo4/SMSApp/MainActivityHelper.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    };
+
+    private static MainActivityState state;
+
+    public static  MainActivityState getState() {
+        return state;
+    }
+
+    public static void setState(MainActivityState state) {
+        state = state;
```
[the remark this answers is not in the document]
Comment (MarcoTomma) [reply app/src/main/java/com/gruppo4/SMSApp/MainActivityHelper.java]: It's a static context so I can't access it with `this.*`...so I'll change the name

File: app/src/main/java/com/gruppo4/SMSApp/MessageReceivedService.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+     * Two actions available in the notification represented by two intents:
+     * Stop --> stops the ringtone
+     * Open --> opens the MainActivity
+     *
+     * @return  id of the notification just created
+     */
+    private int createNotification() {
+
+        //Creating unique id
+        int notif_id = (int) System.currentTimeMillis();
```
[the remark this answers is not in the document]
Comment (MarcoTomma) [reply app/src/main/java/com/gruppo4/SMSApp/MessageReceivedService.java]: Thank you!

File: sms/src/main/java/com/gruppo4/sms/dataLink/background/SMSKeepAliveService.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        broadcastReceiverInstance = null;
+    }
+
+    /**
+     * Registers the broadcastReceiver to this service
+     */
+    private void registerMessageBroadcastReceiver() {
+        broadcastReceiverInstance = new SMSReceivedBroadcastReceiver();
+        IntentFilter filter = new IntentFilter("android.provider.Telephony.SMS_RECEIVED");
+        filter.setPriority(100);
```
[the remark this answers is not in the document]
Comment (MarcoTomma) [reply sms/src/main/java/com/gruppo4/sms/dataLink/background/SMSKeepAliveService.java]: I think this is not countend as hardwiring but I can add a comment instead to make it more readable

File: sms/src/main/java/com/gruppo4/sms/dataLink/background/SMSKeepAliveService.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @author Marco Tommasini
+ * @author Luca Crema
+ */
+public class SMSKeepAliveService extends Service {
+
+    private static SMSReceivedBroadcastReceiver broadcastReceiverInstance;
+
+    @Override
+    public IBinder onBind(Intent arg0) {
```
[the remark this answers is not in the document]
Comment (MarcoTomma) [reply sms/src/main/java/com/gruppo4/sms/dataLink/background/SMSKeepAliveService.java]: I'll add some explanation for making more readable like you suggested

**Review by MarcoTomma — COMMENTED** [review]
Thank you so much for the effort you've done to catch every little missing detail, I'll fix everything!
Don't worry about your first comment, is not false what you said but the point was to prepare a piece of code that I was proud of and it is what I've done.
Thank you again and have a good day!

File: app/src/main/java/com/gruppo4/SMSApp/MainActivityHelper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tonin-alessandra) [reply app/src/main/java/com/gruppo4/SMSApp/MainActivityHelper.java]: Oh you're right, sorry. So yes, changing the name is a good idea.

**Pull request closed without merging**